taichi/transforms/type_check.cpp

```diff
@@ -553,6 +553,26 @@ class TypeCheck : public IRVisitor {
     stmt->ret_type = stmt->var->ret_type;
     stmt->ret_type.set_is_pointer(true);
   }
+
+  void visit(MatrixInitStmt *stmt) override {
+    TI_ASSERT_INFO(stmt->ret_type->is<TensorType>(),
+                   "Matrix should have tensor type, got {}",
+                   stmt->ret_type->to_string());
+    auto tensor_type = stmt->ret_type->as<TensorType>();
+    auto element_dtype = tensor_type->get_element_type();
+    for (auto elt : stmt->values) {
+      element_dtype = promoted_type(element_dtype, elt->ret_type);
+    }
+    for (int i = 0; i < stmt->values.size(); ++i) {
+      if (element_dtype != stmt->values[i]->ret_type) {
+        cast(stmt->values[i], element_dtype);
```

Comment:
Just complaining: 

I know this get fused with CHI IR Passes thus cause no side-effects. However, it is a bit stupid to initialize the constant values with a "default type" then cast it to the dtype specified by the Matrix (this also happens to some other data structures). 

In the future we might want to directly initialize the constants to it's specified dtype.

Reply:
Agree. One possible solution is to enable parent(s) AST nodes to inform their children about how to initialize the most general parts (i.e. fields defined in `Stmt`/`Expression`) when constructing the AST.

Reply:
In the previous implementation, `dt` is an optional argument to `make_matrix` and will keep as is when passed to FrontendIR. I remember I changed to automatically infer it using some utilities already implemented.

Reply:
Then shall we remove the part handling `unknown` in `type_check.cpp`?